java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java

```diff
@@ -361,6 +381,170 @@ public void testEmptyStreamInStreamingIPC() throws IOException {
 
   }
 
+  @Test
+  public void testDictionaryReplacement() throws Exception {
+    VarCharVector vector1 = newVarCharVector("varchar1", allocator);
+    setVector(vector1,
+        "foo".getBytes(StandardCharsets.UTF_8),
+        "bar".getBytes(StandardCharsets.UTF_8),
+        "baz".getBytes(StandardCharsets.UTF_8),
+        "bar".getBytes(StandardCharsets.UTF_8));
+
+    FieldVector encodedVector1 = (FieldVector) DictionaryEncoder.encode(vector1, dictionary1);
+
+    VarCharVector vector2 = newVarCharVector("varchar2", allocator);
+    setVector(vector2,
+        "foo".getBytes(StandardCharsets.UTF_8),
+        "foo".getBytes(StandardCharsets.UTF_8),
+        "foo".getBytes(StandardCharsets.UTF_8),
+        "foo".getBytes(StandardCharsets.UTF_8));
+
+    FieldVector encodedVector2 = (FieldVector) DictionaryEncoder.encode(vector2, dictionary1);
+
+    DictionaryProvider.MapDictionaryProvider provider = new DictionaryProvider.MapDictionaryProvider();
+    provider.put(dictionary1);
+    List<Field> schemaFields = new ArrayList<>();
+    schemaFields.add(DictionaryUtility.toMessageFormat(encodedVector1.getField(), provider, new HashSet<>()));
+    schemaFields.add(DictionaryUtility.toMessageFormat(encodedVector2.getField(), provider, new HashSet<>()));
+    Schema schema = new Schema(schemaFields);
+
+    ByteArrayOutputStream outStream = new ByteArrayOutputStream();
+    WriteChannel out = new WriteChannel(newChannel(outStream));
+
+    // write schema
+    MessageSerializer.serialize(out, schema);
+
+    List<AutoCloseable> closeableList = new ArrayList<>();
+
+    // write non-delta dictionary with id=1
+    serializeDictionaryBatch(out, dictionary3, false, closeableList);
+
+    // write non-delta dictionary with id=1
+    serializeDictionaryBatch(out, dictionary1, false, closeableList);
+
+    // write recordBatch2
+    serializeRecordBatch(out, Arrays.asList(encodedVector1, encodedVector2), closeableList);
```

Comment:
is this necessary for this test?

Reply:
I guess yes, if we only sent dictionaries without batches, seems dictionaries won't be read since we made this change in https://issues.apache.org/jira/browse/ARROW-6040